File: samples/Delegates/delegates.c
typedef struct MonoVTable MonoVTable;

typedef struct _MonoThreadsSync MonoThreadsSync;

typedef struct {
	MonoVTable *vtable;
	MonoThreadsSync *synchronisation;
} MonoObject;

typedef struct _MonoDelegate
{
	MonoObject object;
	MonoObject *target_type;
	MonoObject *target;
	void *method_name;
	void *method_ptr;
	/* TODO: complete stuct
	/* void *delegate_trampoline; */
	/* void *method_info; */
} MonoDelegate;

typedef struct _MonoMulticastDelegate MonoMulticastDelegate;
struct _MonoMulticastDelegate
{
	MonoDelegate d;
	MonoMulticastDelegate *prev;
};

typedef struct _Number$Display
{
	MonoMulticastDelegate md;
} Number$Display;

typedef struct _Number Number;
struct _Number
{
    MonoObject object;
	unsigned short val;
	Number *pred;
	Number *succ;
};

typedef struct _Display
{
	MonoObject object;
	unsigned short count;
} Display;

void
delegate_ctor (MonoDelegate *d, MonoObject *target, void *method_ptr)
{
    d->target = target;
	d->method_ptr = method_ptr;
}

unsigned short
Number$Display$Invoke (Number$Display *obj, Number *v)
{
	if (obj->md.d.target)
	{

		unsigned short (*invoke)(MonoObject* obj, Number* v);

		invoke = (unsigned short (*)(MonoObject* obj, Number* v)) obj->md.d.method_ptr;

		return invoke(obj->md.d.target, v);
	}
	else
	{
		unsigned short (*invoke)(Number* v);

		invoke = (unsigned short (*)(Number* v)) obj->md.d.method_ptr;

		return invoke(v);
	}
}

void
Number$ctor (Number *obj, unsigned short val, Number *p, Number *s)
{
	obj->val = val;
	obj->pred = p;
	obj->succ = s;
}

unsigned short
Number$Value (Number *obj)
{
	return obj->val;
}

Number*
Number$Predecessor (Number *obj)
{
	return obj->pred;
}

Number*
Number$Successor (Number *obj)
{
	return obj->succ;
}

unsigned short
Number$show (Number *obj, Number$Display *showNumbers)
{
	return Number$Display$Invoke (showNumbers, obj);
}

void
Display$show (unsigned short val)
{
	printf ("%d\n", val);
}

unsigned short
Display$showPredecessor (Display *obj, Number *num)
{
	 Display$show (Number$Value (Number$Predecessor (num)));

    ++obj->count;

	return obj->count;
}

unsigned short
Display$showSuccessor (Display *obj, Number *num)
{
    Display$show (Number$Value (Number$Successor (num)));

    ++obj->count;

	return obj->count;
}

unsigned short
Display$sshow (Number *num)
{
    Display$show (Number$Value (num));

	return 1;
}

int
main ()
{
	Number four, five, six;

	Number$Display staticShow, showPred, showSucc;

	Display display;
	display.count = 0;

	Number$ctor (&four, 4, 0, 0);
	Number$ctor (&six, 6, 0, 0);
	Number$ctor (&five, 5, &four, &six);

	delegate_ctor ((MonoDelegate*)&staticShow, 0, Display$sshow);
	delegate_ctor ((MonoDelegate*)&showPred, (MonoObject*)&display, Display$showPredecessor);
	delegate_ctor ((MonoDelegate*)&showSucc, (MonoObject*)&display, Display$showSuccessor);

	Display$show (Number$show (&five, &staticShow));
	Display$show (Number$show (&five, &showPred));
	Display$show (Number$show (&five, &showSucc));
    
	return 0;    
}

File: samples/Delegates/Delegates.cs
using System;

namespace Delegates
{	
	class Number
	{
		public delegate ushort Display(Number v);
        
		private ushort val;
		private Number pred;
		private Number succ;

		public ushort Value
		{
			get {return this.val;}
		}

		public Number Predecessor
		{
			get
			{
				return this.pred;
			}
		}			

		public Number Successor
		{
			get
			{
				return this.succ;
			}
		}			


		public Number (ushort v, Number p, Number s)
		{
			this.val = v;

			this.pred = p;
			
			this.succ = s;
		}

		public ushort show(Number.Display showNumbers)
		{
			return showNumbers(this);
		}
	}
	

	class Display
	{
		protected static ushort scount = 0;
		
		protected ushort count = 0;		

		public ushort showNumber(Number num)
		{			
			show(num.Value);
            
			++count;
                        			
			return this.count;
		}

		public virtual ushort showPredecessor(Number num)
		{
			show(num.Predecessor.Value);

			++count;
                        			
			return this.count;
		}

		public virtual ushort showSuccessor(Number num)
		{
			show(num.Successor.Value);

			++count;
                        			
			return this.count;
		}


		public static ushort sshow(Number num)
		{
			show(num.Value);

			++scount;
                        			
			return scount;
		}
		
		
		public static void show(ushort val)
		{
			brickOS.conio.cputw (val);
			brickOS.unistd.sleep(1);
			brickOS.conio.cls();
			brickOS.unistd.msleep(500);
						
			//Console.WriteLine(val);
		}
	}

	class SpecialDisplay : Display
	{
		public override ushort showPredecessor(Number num)
		{
			show(0);

			++count;
                        			
			return this.count;
		}

		public override ushort showSuccessor(Number num)
		{
			show(0xFFFF);

			++count;
                        			
			return this.count;
		}
	}
	
	class Delegates
	{		
		static void Main(string[] args)
		{
			
			Number four = new Number(4, null, null);
			Number six = new Number(6, null, null);
			Number five = new Number(5, four, six);
			
			Number.Display staticShow= new Number.Display(Display.sshow);			

			Display display = new Display();

			Number.Display showPrec = new Number.Display(display.showPredecessor);
			Number.Display showSucc = new Number.Display(display.showSuccessor);

			Display.show(five.show(staticShow));
			Display.show(five.show(showPrec));
			Display.show(five.show(showSucc));

			Display sdisplay = new SpecialDisplay();
			
			showPrec = new Number.Display(sdisplay.showPredecessor);
			showSucc = new Number.Display(sdisplay.showSuccessor);
		
			Display.show(five.show(showPrec));
			Display.show(five.show(showSucc));
			
			//Console.ReadLine();
		}
	}
}